File: modules/student-experience/src/main/java/com/campusos/portal/service/InboxStates.java
package com.campusos.portal.service;

import com.campusos.portal.domain.Actor;
import com.campusos.portal.domain.RequestState;
import com.campusos.portal.domain.RequestType;
import com.campusos.portal.engine.Transition;
import com.campusos.portal.engine.TransitionMatrix;
import com.campusos.portal.engine.WorkflowSpec;
import java.util.Collection;
import java.util.EnumMap;
import java.util.EnumSet;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Set;

/**
 * WHICH STATES ARE WAITING ON WHICH ACTOR — derived from TransitionMatrix at class-init,
 * never hand-written. If an edge is added, moved or removed in the matrix, the inbox follows
 * it on the next start-up. A hand-maintained list here would be a second source of truth for
 * the workflow, which is exactly what the engine exists to prevent.
 */
public final class InboxStates {

    private InboxStates() {}

    private static final Map<Actor, Set<RequestState>> BY_ACTOR = build();

    private static Map<Actor, Set<RequestState>> build() {
        Map<Actor, Set<RequestState>> out = new EnumMap<>(Actor.class);
        for (RequestType type : RequestType.values()) {
            WorkflowSpec spec = TransitionMatrix.spec(type);
            for (Map.Entry<RequestState, java.util.List<Transition>> e : spec.edges().entrySet()) {
                for (Transition t : e.getValue()) {
                    if (t.actor() == Actor.SYSTEM || t.actor() == Actor.STUDENT) continue;
                    out.computeIfAbsent(t.actor(), a -> EnumSet.noneOf(RequestState.class))
                            .add(e.getKey());
                }
            }
        }
        return Map.copyOf(out);
    }

    /** The states one Actor is the human decision-maker for. */
    public static Set<RequestState> awaiting(Actor a) {
        return BY_ACTOR.getOrDefault(a, Set.of());
    }

    /** The union over every role a staff member holds — their whole inbox. */
    public static Set<RequestState> awaiting(Collection<Actor> actors) {
        Set<RequestState> out = new LinkedHashSet<>();
        for (Actor a : actors) out.addAll(awaiting(a));
        return out;
    }
}

File: modules/student-experience-js/src/service/InboxStates.js
import { Actor, RequestType } from '../domain/enums.js';
import { TransitionMatrix } from '../engine/TransitionMatrix.js';

/**
 * WHICH STATES ARE WAITING ON WHICH ACTOR — derived from TransitionMatrix at module-load,
 * never hand-written. If an edge is added, moved or removed in the matrix, the inbox follows
 * it on the next start-up.
 */
function build() {
  const out = {};
  for (const type of Object.values(RequestType)) {
    const spec = TransitionMatrix.spec(type);
    for (const [state, transitions] of Object.entries(spec.edges)) {
      for (const t of transitions) {
        if (t.actor === Actor.SYSTEM || t.actor === Actor.STUDENT) continue;
        (out[t.actor] ??= new Set()).add(state);
      }
    }
  }
  return out;
}

const BY_ACTOR = build();

export const InboxStates = {
  /** The states one Actor is the human decision-maker for. */
  awaitingOne(actor) { return BY_ACTOR[actor] ? [...BY_ACTOR[actor]] : []; },

  /** The union over every role a staff member holds — their whole inbox. */
  awaiting(actors) {
    const out = new Set();
    for (const a of actors) for (const s of this.awaitingOne(a)) out.add(s);
    return [...out];
  },
};
